client/canonical_eventlog.go

```diff
@@ -0,0 +1,266 @@
+package client
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/google/go-tpm/tpm2"
+)
+
+var hashAlgoMapping = map[crypto.Hash]tpm2.Algorithm{
+	crypto.SHA1:   tpm2.AlgSHA1,
+	crypto.SHA256: tpm2.AlgSHA256,
+	crypto.SHA384: tpm2.AlgSHA384,
+	crypto.SHA512: tpm2.AlgSHA512,
+}
+
+const (
+	// CEL spec 5.1
+	recNumTagValue uint8 = 0
+	pcrTagValue    uint8 = 1
+	// nvIndexTagValue uint8 = 2 // nvindex field is not supported
+	digestsTagValue uint8 = 3
+
+	tlvTagFieldLength    int = 1
+	tlvLengthFieldLength int = 4
+
+	recNumValueLength uint32 = 8 // support up to 2^64 records
+	pcrValueLength    uint32 = 1 // support up to 256 PCRs
+)
+
+// TLV definition according to CEL spec, page 16
+type TLV struct {
+	Tag    uint8
+	Length uint32 // big-endian when encoding
+	Value  []byte // with size of Length
+}
+
+// TLVMarshal marhsal a TLV to a byte slice
+func TLVMarshal(tlv TLV) []byte {
+	buf := make([]byte, tlv.Length+uint32(tlvTagFieldLength)+uint32(tlvLengthFieldLength))
+
+	buf[0] = tlv.Tag
+	binary.BigEndian.PutUint32(buf[tlvTagFieldLength:], tlv.Length)
+	copy(buf[tlvTagFieldLength+tlvLengthFieldLength:], tlv.Value)
+
+	return buf
+}
+
+// TLVUnmarshal reads and parse the first TLV from the bytes buffer.
+func TLVUnmarshal(buf *bytes.Buffer) (TLV, error) {
+	var tlv TLV
+	var err error
+
+	// read tag
+	tlv.Tag, err = buf.ReadByte()
+	if err != nil {
+		return TLV{}, err
+	}
+
+	// read length
+	b := make([]byte, tlvLengthFieldLength)
+	l, err := buf.Read(b)
+
+	if err != nil || l < tlvLengthFieldLength {
+		return TLV{}, io.EOF
+	}
+	tlv.Length = binary.BigEndian.Uint32(b)
+
+	// read value
+	tlv.Value = make([]byte, tlv.Length)
+	l, err = buf.Read(tlv.Value)
+	if err != nil || l < int(tlv.Length) {
+		return TLV{}, io.EOF
+	}
+	return tlv, nil
+}
+
+// CELR represents a Canoical Eventlog Record
+type CELR struct {
+	RecNum  TLV
+	PCR     TLV
+	Digests TLV
+	Content TLV
+}
+
+// CELREventData is a interface for the content in CELR.
+type CELREventData interface {
+	GenerateDigest(crypto.Hash) ([]byte, error)
+	GetTLV() TLV
+}
+
+// CEL represents a Canonical Eventlog, which contains a list of CELR
+type CEL struct {
+	Records []CELR
+}
+
+// AppendEvent appends a new record to the CEL.
+func (c *CEL) AppendEvent(tpm io.ReadWriteCloser, pcr int, hashAlgos []crypto.Hash, payload CELREventData) error {
+	digestsMap := make(map[tpm2.Algorithm][]byte)
+	for _, hashAlgo := range hashAlgos {
+		digest, err := payload.GenerateDigest(hashAlgo)
+		if err != nil {
+			return err
+		}
+		digestsMap[hashAlgoMapping[hashAlgo]] = digest
+
+		// TODO: extend the digest to TPM PCR
```

Comment:
it probably shouldn't be the CEL struct's business to extend the PCR. we can have a utility wrapper outside of CEL in `client`. WDYT?

Reply:
I want to make them tightly coupled together, because I thought it doesn't make sense to create an event log without extending to a PCR.
It is also easier to extend to PCR here as the digests are generated here.

Reply:
I think having the "writing the log" operation coupled with the "extending the PCR" operation is fine.